parsons/empower/empower.py

```diff
@@ -0,0 +1,261 @@
+
+        `Args:`
+            api_key: str
+                The Empower provided API key.The Empower provided Client UUID. Not
+                required if ``EMPOWER_API_KEY`` env variable set.
+            empower_uri: str
+                The URI to access the Empower API. The default is currently set to
+                https://api.getempower.com/v1/export. You can set an ``EMPOWER_URI`` env
+                variable or use this URI parameter if a different endpoint is necessary.
+            cache: boolean
+                The Empower API returns all account data after each call. Setting cache
+                to ``True`` stores the blob and then extracts Parsons tables for each method.
+                Setting cache to ``False`` will download all account data for each method call.
+    """
+
+    def __init__(self, api_key=None, empower_uri=None, cache=True):
+        self.api_key = check_env.check("EMPOWER_API_KEY", api_key)
+        self.empower_uri = (
+            check_env.check("EMPOWER_URI", empower_uri, optional=True) or EMPOWER_API_ENDPOINT
+        )
+        self.headers = {"accept": "application/json", "secret-token": self.api_key}
+        self.client = APIConnector(
+            self.empower_uri,
+            headers=self.headers,
+        )
+        self.data = None
+        self.data = self._get_data(cache)
+
+    def _get_data(self, cache):
+        """
+        Gets fresh data from Empower API based on cache setting.
+        """
+
+        if not cache or self.data is None:
+            r = self.client.get_request(self.empower_uri)
+            logger.info("Empower data downloaded.")
+            return r
+
+        else:
+            return self.data
+
+    def _unix_convert(self, ts):
+        """
+        Converts UNIX timestamps to readable timestamps.
+        """
+
+        ts = datetime.utcfromtimestamp(int(ts) / 1000)
+        ts = ts.strftime("%Y-%m-%d %H:%M:%S UTC")
+        return ts
+
+    def _empty_obj(self, obj_name):
+        """
+        Determine if a dict object is empty.
+        """
+
+        if len(self.data[obj_name]) == 0:
+            return True
+        else:
+            return False
+
+    def get_profiles(self):
+        """
+        Get Empower profiles.
+
+        `Returns:`
+            Parsons Table
+                See :ref:`parsons-table` for output options.
+        """
+
+        tbl = Table(self.data["profiles"])
+        for col in ["createdMts", "lastUsedEmpowerMts", "updatedMts"]:
+            tbl.convert_column(col, lambda x: self._unix_convert(x))
+        tbl.remove_column("activeCtaIds")  # Get as a method via get_profiles_active_ctas
+        return tbl
+
+    def get_profiles_active_ctas(self):
+        """
+        Get active ctas assigned to Empower profiles.
+
+        `Returns:`
+            Parsons Table
+                See :ref:`parsons-table` for output options.
+        """
+
+        tbl = Table(self.data["profiles"]).long_table("eid", "activeCtaIds")
```

Comment:
This is pure curiosity, what is a "long table"? Guess that's a part of Parsons core I've never had to touch before

Reply:
You have your identifier columns (the values that tell you what row the data would belong to), then all your other variables are melting into `variable` and `value` columns. So for person A with variables X, Y, Z in a normal table, the long table will have 3 rows: 

```
Person     Variable     Value
A              X                x_value
A              Y                y_value
A              Z                z_value
```

Reply:
Thanks for the explanation!